src/storage-client/src/controller.rs

```diff
@@ -1258,83 +1313,151 @@ where
+            panic!("cannot cancel prepared export: {:?}", err);
+        }
     }
 
     async fn create_exports(
         &mut self,
-        exports: Vec<(CreateExportToken, ExportDescription<Self::Timestamp>)>,
+        exports: Vec<(
+            CreateExportToken<Self::Timestamp>,
+            ExportDescription<Self::Timestamp>,
+        )>,
     ) -> Result<(), StorageError> {
         // Validate first, to avoid corrupting state.
         let mut dedup_hashmap = BTreeMap::<&_, &_>::new();
-        for (CreateExportToken { id, from_id }, desc) in exports.iter() {
+        for (
+            CreateExportToken {
+                id,
+                from_id,
+                acquired_since: _,
+            },
+            desc,
+        ) in exports.iter()
+        {
             if dedup_hashmap.insert(id, desc).is_some() {
-                return Err(StorageError::SourceIdReused(*id));
+                return Err(StorageError::SinkIdReused(*id));
             }
             if let Ok(export) = self.export(*id) {
                 if &export.description != desc {
-                    return Err(StorageError::SourceIdReused(*id));
+                    return Err(StorageError::SinkIdReused(*id));
                 }
             }
             if desc.sink.from != *from_id {
-                return Err(StorageError::SourceIdReused(*id));
-            }
-            if self
-                .state
-                .exported_collections
-                .get(from_id)
-                // Internal logic error NOT due to export not existing
-                .expect("Dangling exported collection")
-                .iter()
-                .find(|from_export_id| *from_export_id == id)
-                .is_none()
-            {
-                return Err(StorageError::SourceIdReused(*id));
+                return Err(StorageError::InvalidUsage(format!(
+                    "sink {id} was prepared using from_id {from_id}, \
+                    but is now presented with from_id {}",
+                    desc.sink.from
+                )));
             }
         }
 
-        for (CreateExportToken { id, from_id }, description) in exports {
-            self.state
-                .exports
-                .insert(id, ExportState::new(description.clone()));
+        for (
+            CreateExportToken {
+                id,
+                from_id,
+                acquired_since,
+            },
+            description,
+        ) in exports
+        {
+            // TODO(aljoscha): I _think_ this description is subtly wrong. But
+            // not completely sure.
+            //
+            // It's worth adding a quick note on write frontiers here.
+            //
+            // The write frontier that sinks communicate back to the controller
+            // indicates that all further writes will happen at a time `t` such
+            // that `!timely::ParitalOrder::less_than(&t, &write_frontier)` is
```

Comment:
```suggestion
            // that `!timely::PartialOrder::less_than(&t, &write_frontier)` is
```

But also there's a type error here... we're comparing a frontier to a timestamp.

Reply:
I'm reserving changes to this block of comments and how the sink does it's upper reporting to follow-up work. Hoping to get the rest of this merged rather soon to get Sean unblocked.

Reply:
That seems fine, especially since this comment is simply relocated.